libnavigation-core/src/main/java/com/mapbox/navigation/core/fasterroute/FasterRouteController.kt

```diff
@@ -0,0 +1,70 @@
+package com.mapbox.navigation.core.fasterroute
+
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.api.directions.v5.models.RouteOptions
+import com.mapbox.navigation.base.extensions.ifNonNull
+import com.mapbox.navigation.core.directions.session.DirectionsSession
+import com.mapbox.navigation.core.directions.session.RoutesRequestCallback
+import com.mapbox.navigation.core.trip.session.TripSession
+import com.mapbox.navigation.utils.timer.MapboxTimer
+import java.util.concurrent.CopyOnWriteArrayList
+
+internal class FasterRouteController(
+    private val directionsSession: DirectionsSession,
+    private val tripSession: TripSession
```

Comment:
Do you think both should be `WeakReference`? How are we going to access this from `UI SDK`? 
We would have to do the following:

`NavigationViewModel.java`
```
val fasterRouteController = FasterRouteController(mapboxNavigation.getDirectionSession(), mapboxNavigation.getTripSession());
```

We cannot do `mapboxNavigation.getDirectionSession()` and `mapboxNavigation.getTripSession()` because `MapboxNavigation` has no such methods.

Reply:
The lifecycle will be the same as these. So it shouldn't need a WeakReference
https://github.com/mapbox/mapbox-navigation-android/blob/master/libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt#L136-L138

this class is `internal` so it shouldn't have the issue leaking. It gives access to `UI SDK` with the call back (which we need to unsubscribe from or there will be a leak)

^ though, none of this is changed from this refactor

Reply:
This still doesn't answer my question as to how does the UI SDK create an instance of this class?

Reply:
Oh sorry maybe didn't catch it exactly. UI will not create an instance of this class. Only MapboxNavigation
https://github.com/mapbox/mapbox-navigation-android/pull/2577/files#diff-e3ca7b557efaeed833ca6f8f76dc2d03R193

Reply:
Makes sense.